frame/contracts/src/wasm/runtime.rs

```diff
@@ -571,6 +586,30 @@ define_env!(Env, <E: Ext>,
 		}
 	},
 
+	// Remove the calling account and transfer remaining balance.
+	//
+	// This function never returns. Either the termination was successful and the
+	// execution of the destroyed contract is halted. Or it failed during the termination
+	// which is considered fatal and results in a trap + rollback.
+	//
+	// - beneficiary_ptr: a pointer to the address of the beneficiary account where all
+	//   where all remaining funds of the caller are transfered.
+	//   Should be decodable as an `T::AccountId`. Traps otherwise.
+	// - beneficiary_len: length of the address buffer.
+	ext_terminate(
+		ctx,
+		beneficiary_ptr: u32,
+		beneficiary_len: u32
+	) => {
+		let beneficiary: <<E as Ext>::T as frame_system::Trait>::AccountId =
+			read_sandbox_memory_as(ctx, beneficiary_ptr, beneficiary_len)?;
+
+		if let Ok(_) = ctx.ext.terminate(&beneficiary, ctx.gas_meter) {
+			ctx.special_trap = Some(SpecialTrap::Termination);
+		}
+		Err(sp_sandbox::HostError)
```

Comment:
This function always returns an error?

Reply:
No: It always terminates the execution of the current call (this is what the return of `Err` does). But in the case of a successful `terminate` the `special_trap` is set. This trap signals to the VM that the termination of the call was successful: No revert takes place and success is signaled to the caller.

The debatable reasoning behind that is that `ext_terminate` should never return. Neither on success nor on failure. Returning to a terminated contract feels wrong to me. Returning on failure when we doesn't on success feels like a recipe for disaster. I am willing to be swayed, though.